### ISSUE TITLE
[FastPitch/Pytorch] impact of global batch size on loss

### ISSUE THREAD

**Issue by dsplog** [NVIDIA/DeepLearningExamples#1001]
Related to **( FastPitch/PyTorch )** 

**Describe the bug**
This is not a bug, but a clarification request on the expected behaviour on global batchsize. 
When training on a smaller gpu with 6GB ram, had to use num_gpus=1, batch_size = 8
* when inadvertantly checked with grad_accumulation = 2, could see that the training loss is decreasing rapdily. 
at the end of epoch1, training/validation loss reduced to 4.64 !!

* when checked with grad_accumulation = 32, as recommended to make the overall number 256, then the loss is reducing only to ~34 after epoch1

**To Reproduce**

* grad_accumulation = 2
```
(pytorch_gpu) home@home-desktop:/home/home/DeepLearningExamples/PyTorch/SpeechSynthesis/FastPitch$ bash scripts/train.sh 

WARNING: Global batch size changed from 256 to 16.

AMP=false, 1x8x2 (global batch size 16)
<snip> 
DLL 2021-09-10 11:52:23.451233 - epoch    1 | iter 780/781 | loss  4.58 | mel loss  0.89 | kl loss 0.00969 | kl weight 0.01000 |   12427.52 frames/s | took 0.74 s | lrate 2.47e-03
DLL 2021-09-10 11:52:24.157096 - epoch    1 | iter 781/781 | loss  4.64 | mel loss  0.95 | kl loss 0.01064 | kl weight 0.01000 |   11136.29 frames/s | took 0.70 s | lrate 2.47e-03
DLL 2021-09-10 11:52:31.234292 - epoch    1 |   avg val loss  4.80 |   avg val mel loss  0.93 |     310.23 frames/s | took 7.07 s

```

* grad_accumulation = 32
```
(pytorch_gpu) home@home-desktop:/home/home/DeepLearningExamples/PyTorch/SpeechSynthesis/FastPitch$ bash scripts/train.sh 
AMP=false, 1x8x32 (global batch size 256)
<snip>
DLL 2021-09-10 12:08:38.223088 - epoch    1 | iter  48/48 | loss 34.78 | mel loss 28.87 | kl loss 0.02270 | kl weight 0.01000 |   10720.52 frames/s | took 13.65 s | lrate 1.52e-04
DLL 2021-09-10 12:08:38.224007 - epoch    1 | avg train loss 39.76 | avg train mel loss 32.85 |   10811.11 frames/s | took 644.21 s
DLL 2021-09-10 12:08:48.008687 - epoch    1 |   avg val loss 34.58 |   avg val mel loss 28.88 |     224.34 frames/s | took 9.78 s

```

**Expected behavior**
based on above, have a suspicion using grad_accumulation=2 is falsely showing faster convergence. as per your data, the final loss is ~2.6 after ~1200 epochs.  can you plz help clarify. 

**Environment**
Please provide at least:
* Container version (e.g. pytorch:1.9.0.+cu111):
* GPUs in the system: (e.g. 1x RTX2060-6GB):
* CUDA driver version (e.g. 470.57.02 ):

**Comment by zsh01** [comment]
Hi @dsplog! Sorry for a late reply.

> This is not a bug, but a clarification request on the expected behaviour on global batchsize.

Keep in mind that there's an inintial LR warmup stage, which takes 1k steps.
When you decrease global batch size but do not adjust warmup steps, you rapidly ramp up LR. You can see it as the last entry in your output -- `2.47e-03` with GA=2 and `1.52e-04` with GA=32.

Typically LR is increased along with batch size. So when you decrease the global batch size and keep LR, the effect is amplified. Increasing the LR too early might lead to worse convergence - the loss might decrease rapidly at first, but stall later on.

**Comment by alancucki** [comment]
Hi @dsplog! Sorry for a late reply.

> This is not a bug, but a clarification request on the expected behaviour on global batchsize.


Keep in mind that there's an inintial LR warmup stage, which takes 1k steps.
When you decrease global batch size but do not adjust warmup steps, you rapidly ramp up LR. You can see it as the last entry in your output -- 2.47e-03 with GA=2 and 1.52e-04 with GA=32.

Typically LR is increased along with batch size. So when you decrease the global batch size and keep LR, the effect is amplified. Increasing the LR too early might lead to worse convergence - the loss might decrease rapidly at first, but stall later on.

**Comment by dsplog** [comment]
ah, learning rate! did not notice the 2e-3 vs 1.5e-4 :-)  thanks for clarification. closing.
